packages/sdk/lib/__tests__/validation.test.ts

```diff
@@ -0,0 +1,220 @@
+      (methodName, methodArgs) => {
+        const validationPlugin = createPlugin();
+        validationPlugin.validate = jest.fn(() => ({ valid: false, message: `${methodName} error` }));
+
+        const plugin = createPlugin();
+
+        const itly = new Itly();
+        itly.load({
+          plugins: [validationPlugin, plugin],
+          validation: Validation.Disabled,
+        });
+
+        callItlyMethod(itly, methodName, methodArgs);
+        expect(validationPlugin.validate).toHaveBeenCalledTimes(0);
+
+        const [pluginMethod, pluginPostMethod] = getPluginMethods(plugin, methodName);
+        expect(pluginMethod).toHaveBeenCalledTimes(1);
+        expect(pluginPostMethod).toHaveBeenCalledTimes(1);
+      });
+  });
+
+  describe('should throw validation error if validate() throws exception and options.validation = ErrorOnInvalid', () => {
+    test.each(trackMethods)('%s',
+      (methodName, methodArgs) => {
+        const validationPlugin = createPlugin();
+        validationPlugin.validate = jest.fn(() => {
+          throw new Error(`${methodName} exception`);
+        });
+
+        const plugin = createPlugin();
+
+        const itly = new Itly();
+        itly.load({
+          plugins: [validationPlugin, plugin],
+          validation: Validation.ErrorOnInvalid,
+        });
+
+        expect(() => callItlyMethod(itly, methodName, methodArgs)).toThrow(`Validation Error: ${methodName} exception`);
+        expect(validationPlugin.validate).toHaveBeenCalledTimes(1);
+
+        const [pluginMethod, pluginPostMethod] = getPluginMethods(plugin, methodName);
+        expect(pluginMethod).toHaveBeenCalledTimes(0);
+        expect(pluginPostMethod).toHaveBeenCalledTimes(1);
+      });
+  });
+
+  describe('should throw validation error if context is not valid and options.validation = ErrorOnInvalid', () => {
+    test.each(trackMethods.filter(([methodName]) => methodName === 'track'))('%s',
+      (methodName, methodArgs) => {
+        const validationPlugin = createPlugin();
+        validationPlugin.validate = jest.fn(
+          (event: Event) => (event.name === 'context' ? { valid: false, message: 'context error' } : { valid: true }),
+        );
+
+        const plugin = createPlugin();
+
+        const itly = new Itly();
+        itly.load({
+          context: {
+            a: 123,
+          },
+          plugins: [validationPlugin, plugin],
+          validation: Validation.ErrorOnInvalid,
+        });
+
+        expect(() => callItlyMethod(itly, methodName, methodArgs)).toThrow('Validation Error: context error');
+        expect(validationPlugin.validate).toHaveBeenCalledTimes(2);
+
+        const [pluginMethod, pluginPostMethod] = getPluginMethods(plugin, methodName);
+        expect(pluginMethod).toHaveBeenCalledTimes(0);
+        expect(pluginPostMethod).toHaveBeenCalledTimes(1);
+      });
+  });
+
+  describe('should track valid event if context is not valid and options.validation = ErrorOnInvalid', () => {
```

Comment:
I found this confusing since it says event/track. I think `should post valid properties if context is not valid and options.validation = ErrorOnInvalid` is a little better.

Reply:
Fixed